File: server/ctc.ts
// Pure CTC helpers — no runtime dependencies, unit-tested against the reference fixture
// (web/scripts/ctc-fixture.json, exported from the original Python pipeline). main.ts wires
// these to onnxruntime-node output.

/** Row-wise softmax over the vocab dimension of a [frames, vocab] logits buffer. */
export function softmaxRows(logits: Float32Array | number[], frames: number, vocab: number): Float32Array {
  const out = new Float32Array(frames * vocab);
  for (let t = 0; t < frames; t++) {
    const base = t * vocab;
    let max = -Infinity;
    for (let v = 0; v < vocab; v++) max = Math.max(max, logits[base + v]);
    let sum = 0;
    for (let v = 0; v < vocab; v++) {
      const e = Math.exp(logits[base + v] - max);
      out[base + v] = e;
      sum += e;
    }
    for (let v = 0; v < vocab; v++) out[base + v] /= sum;
  }
  return out;
}

/** Greedy CTC decode from softmax probs: drop repeats, drop blank, join, delimiter -> space. */
export function greedyDecode(
  probs: Float32Array | number[],
  frames: number,
  vocab: number,
  idToToken: Map<number, string>,
  blank: number,
  delimiter: string,
): string {
  let prev = -1;
  const tokens: string[] = [];
  for (let t = 0; t < frames; t++) {
    const base = t * vocab;
    let best = 0;
    let bestVal = probs[base];
    for (let v = 1; v < vocab; v++) {
      if (probs[base + v] > bestVal) {
        bestVal = probs[base + v];
        best = v;
      }
    }
    if (best !== prev && best !== blank) {
      const tok = idToToken.get(best) ?? '';
      // Skip special tokens ([UNK], [PAD], <s>, </s>) so they don't leak into the transcript.
      if (tok && tok[0] !== '[' && tok[0] !== '<') tokens.push(tok);
    }
    prev = best;
  }
  return tokens.join('').split(delimiter).join(' ').replace(/\s+/g, ' ').trim();
}

/**
 * CTC Viterbi forced alignment of target ids to frames, returning each target token's mean
 * acoustic posterior (Goodness-of-Pronunciation, 0..1). Extended sequence interleaves blanks:
 * [blank, y0, blank, y1, ..., blank].
 */
export function forcedAlignGop(
  probs: Float32Array | number[],
  frames: number,
  vocab: number,
  ids: number[],
  blank: number,
): number[] {
  if (ids.length === 0 || frames === 0) return [];
  const ext: number[] = [blank];
  for (const id of ids) ext.push(id, blank);
  const S = ext.length;
  const NEG = -1e30;
  const lp = (t: number, tok: number) => Math.log(probs[t * vocab + tok] + 1e-30);

  const dp = new Float64Array(frames * S).fill(NEG);
  const bp = new Int32Array(frames * S);
  dp[0] = lp(0, ext[0]);
  if (S > 1) dp[1] = lp(0, ext[1]);
  for (let t = 1; t < frames; t++) {
    for (let s = 0; s < S; s++) {
      let best = dp[(t - 1) * S + s];
      let arg = s;
      if (s > 0 && dp[(t - 1) * S + s - 1] > best) {
        best = dp[(t - 1) * S + s - 1];
        arg = s - 1;
      }
      if (s > 1 && ext[s] !== blank && ext[s] !== ext[s - 2] && dp[(t - 1) * S + s - 2] > best) {
        best = dp[(t - 1) * S + s - 2];
        arg = s - 2;
      }
      dp[t * S + s] = lp(t, ext[s]) + best;
      bp[t * S + s] = arg;
    }
  }
  let s = dp[(frames - 1) * S + S - 1] >= dp[(frames - 1) * S + S - 2] ? S - 1 : S - 2;
  const align = new Int32Array(frames);
  for (let t = frames - 1; t >= 0; t--) {
    align[t] = s;
    s = bp[t * S + s];
  }
  const scores: number[] = [];
  for (let j = 0; j < ids.length; j++) {
    const es = 2 * j + 1;
    let sum = 0;
    let n = 0;
    for (let t = 0; t < frames; t++) {
      if (align[t] === es) {
        sum += probs[t * vocab + ids[j]];
        n++;
      }
    }
    scores.push(n > 0 ? sum / n : 0);
  }
  return scores;
}

File: server/ctc.py
"""Pure CTC math for Kyrgyz ASR: softmax, greedy decode, Viterbi forced-alignment GOP.

Mirrors web/src/lib/ctc.ts (which it replaced as the runtime implementation) and is locked to
the same reference fixture — web/scripts/ctc-fixture.json — by test_ctc.py.
"""
from __future__ import annotations

import numpy as np


def softmax_rows(logits: np.ndarray) -> np.ndarray:
    """Row-wise softmax over the vocab dimension of a [frames, vocab] array."""
    z = logits - logits.max(-1, keepdims=True)
    e = np.exp(z)
    return e / e.sum(-1, keepdims=True)


def greedy_decode(probs: np.ndarray, id_to_token: dict[int, str], blank: int, delimiter: str) -> str:
    """Greedy CTC decode: drop repeats, drop blank, skip special tokens, delimiter -> space."""
    ids = probs.argmax(-1)
    prev = -1
    out: list[str] = []
    for i in ids:
        i = int(i)
        if i != prev and i != blank:
            tok = id_to_token.get(i, "")
            if tok and tok[0] not in "[<":
                out.append(tok)
        prev = i
    return " ".join("".join(out).replace(delimiter, " ").split())


def forced_align_gop(probs: np.ndarray, ids: list[int], blank: int) -> list[float]:
    """CTC Viterbi forced alignment of target ids; per-token mean acoustic posterior (0..1).

    Extended sequence interleaves blanks: [blank, y0, blank, y1, ..., blank].
    """
    frames = probs.shape[0]
    if not ids or frames == 0:
        return []
    ext = [blank]
    for t in ids:
        ext += [t, blank]
    S = len(ext)
    NEG = -1e30
    logp = np.log(probs + 1e-30)

    dp = np.full((frames, S), NEG)
    bp = np.zeros((frames, S), np.int32)
    dp[0, 0] = logp[0, ext[0]]
    if S > 1:
        dp[0, 1] = logp[0, ext[1]]
    for t in range(1, frames):
        for s in range(S):
            best, arg = dp[t - 1, s], s
            if s > 0 and dp[t - 1, s - 1] > best:
                best, arg = dp[t - 1, s - 1], s - 1
            if s > 1 and ext[s] != blank and ext[s] != ext[s - 2] and dp[t - 1, s - 2] > best:
                best, arg = dp[t - 1, s - 2], s - 2
            dp[t, s] = logp[t, ext[s]] + best
            bp[t, s] = arg
    s = S - 1 if dp[frames - 1, S - 1] >= dp[frames - 1, S - 2] else S - 2
    align = np.zeros(frames, np.int32)
    for t in range(frames - 1, -1, -1):
        align[t] = s
        s = bp[t, s]

    scores: list[float] = []
    for j, tok in enumerate(ids):
        mask = align == 2 * j + 1
        scores.append(float(probs[mask, tok].mean()) if mask.any() else 0.0)
    return scores
